File: js/question.js
var readline = require('readline');

var getRL = function(){
	return readline.createInterface({
	  input: process.stdin,
	  output: process.stdout
	});
}
;
var multi = function(string, options){

	var rl = getRL();

	var question = string + '\n';

	var callbackMap = {};

	options.forEach(function(option){
		question += option.code;
		question += ' - ';
		question += option.description;
		question += '\n';

		callbackMap[option.code] = option.callback;
	});

	question += '\n';

	rl.question(question, function(answer) {
		rl.close();
	  	if(typeof callbackMap[answer] === 'undefined'){
	  		console.log('sorry, unrecognised response');
	  		multi(string, options);
	  		return;
	  	}

	  	callbackMap[answer]();
	});
};



var ask = function(string, callback){
	var rl = getRL();

	rl.question(string, function(answer){
		rl.close();
		callback(answer);
	});
};

exports.multi = multi;
exports.ask = ask;

File: java/cli/Question.java
package cli;

import cli.Answer;
import cli.ReadLine;

public class Question{

	public void ask(String question, Answer[] answers){

		System.out.println(question);

		for(int i = 0; i < answers.length; i++){
			String output = String.format("%s - %s", answers[i].code, answers[i].description);
			System.out.println(output);			
		}

		ReadLine reader = new ReadLine();
		String answer = reader.read();

		for(int i = 0; i < answers.length; i++){
			if(answer.equals(answers[i].code)){
				answers[i].effect.execute();
				return;
			}
		}

		ask(question, answers);

	}
}